src/nvim/os/keyboard.c

```diff
@@ -0,0 +1,231 @@
+    return false;
+  }
+
+  _XDisplay *const display = _XOpenDisplay(NULL);
+  if (display == NULL) {
+    *err = "Couldn't open the display";
+    return false;
+  }
+
+  _XkbStateRec xkb_state;
+  const _XStatus status = _XkbGetState(display, _XkbUseCoreKbd, &xkb_state);
+  if (status) {
+    *err = "Non-zero status returned by XkbGetState";
+    CLOSE_DISPLAY(display);
+  }
+  *mods = xkb_state.data[kLockedModsOffset];
+
+  CLOSE_DISPLAY(display);
+  return true;
+#undef CLOSE_DISPLAY
+}
+
+#elif defined(WIN32)
+/// Get information about Num/Caps Lock state in Windows
+///
+/// To be used in os_mods_status() function.
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a bitmas
+///
+/// @param[out]  err  Location where error message is to be saved, set to NULL
+///                   if no error.
+///
+/// @return true in case of error, false otherwise.
+bool os_get_locks_status(int *const mods, const char **const err)
+{
+  *mods = 0;
+  *err = NULL;
+  return (GetKeyState(VK_CAPITAL) & 0x0001) != 0;
+  if (GetKeyState(VK_CAPITAL) & 0x0001) {
+    *mods = *mods | kCapsLock;
+  }
+  if (GetKeyState(VK_NUMLOCK) & 0x0001) {
+    *mods = *mods | kNumLock;
+  }
+  return true;
+}
+
+#elif defined(__APPLE__)
+/// Get information about Num/Caps Lock state in MacOS
+///
+/// To be used in os_mods_status() function.
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a bitmas
+///
+/// @param[out]  err  Location where error message is to be saved, set to NULL
+///                   if no error.
+///
+/// @return true in case of error, false otherwise.
+static bool os_get_locks_status(int *const mods, const char **const err)
+{
+  *err = NULL;
+  io_service_t ios;
+  io_connect_t ioc;
+  CFMutableDictionaryRef mdict;
+  bool stateCaps, stateNums;
+  mdict = IOServiceMatching(kIOHIDSystemClass);
+  ios = IOServiceGetMatchingService(kIOMasterPortDefault,
+                                    (CFDictionaryRef)mdict);
+  if (!ios) {
+    if (mdict) {
+      CFRelease(mdict);
+    }
+    *err = "Unable to start service to query data";
+    return false;
+  }
+
+  if (IOServiceOpen(ios, mach_task_self(), kIOHIDParamConnectType,
+                    &ioc) != kIOReturnSuccess) {
+    IOObjectRelease(ios);
+    *err = "Unable to query data";
+    return false;
+  }
+  IOObjectRelease(ios);
+
+  if (IOHIDGetModifierLockState(ioc, kIOHIDCapsLockState,
+                                &stateCaps) != kIOReturnSuccess) {
+    *err = "Unable to lock the state";
+    return false;
+  }
+
+  if (IOHIDGetModifierLockState(ioc, kIOHIDCapsLockState,
+                                &stateNums) != kIOReturnSuccess) {
+    *err = "Unable to lock the state";
+    return false;
+  }
+  if (stateCaps) {
+    *mods = *mods | kCapsLock;
+  }
+  if (stateNums) {
+    *mods = *mods | kNumLock;
+  }
+  IOServiceClose(ioc);
+  return true;
+}
```

Comment:
Should have `#else` with a dummy returning true and zero mask I think.

Reply:
OK, sure, I think this would be perfect, whatif there was some other platform using neovim